### ISSUE TITLE
Non-linter based conventions

### ISSUE THREAD

**Issue by maciejhirsz** [styladev/stylaguide#15]
We have configured a set of rules that can be checked by the linter, but there is more to code style than what linter (at least eslint given dynamic nature of JS) can check, mostly dealing with naming and ordering of things in code.

Some things that are pretty common conventions from any OOP-able language out of the top of my head:

- Class (ES5 constructor function) name must begin with an upper case letter, no other identifier can begin with an upper case letter.
- Class body order: properties first (not applicable without stage2 transpiling atm), then constructor, then public methods, then private methods.
- Variables holding a boolean or functions / methods returning a boolean must begin with one of: `is`, `has`, `should`.
    - Good:
      ```js
      shouldComponentUpdate() {
          return this.hasChanged;
      }
      ```
    - Bad:
      ```js
      updateComponent() {
          return this.changed;
      }
      ```

Some of the stuff I listed we already follow quite rigorously, others are hit/miss depending on who wrote the code and when. Is this something we might be interested in adding to the guide?

**Comment by webstrap** [comment]
Yes, sounds good. The linter should be able to check class names, in Java even more than that.

**Comment by janwilamowski** [comment]
I generally agree, but I'm not sure about the ordering of methods. I like to have small private helper methods close by so I don't have to jump up and down in the file (especially when the editor doesn't support that well). Are any of these issues an actual problem that would be solved with adding more restrictions?

**Comment by webstrap** [comment]
I would see it as general guidelines on which you agree, where you can differ when it makes sense. In general i guess people would expect to first see the public "interface" of the class and then private methods, but might make sense in certain situations to have it rather grouped by context (and inside that again public first ;) )

**Comment by maciejhirsz** [comment]
Ordering of public and private methods is tricky since JS doesn't actually have the notion of private methods. We could use `@private` JSDoc marker for that, currently single story page is broken on `dev` because a public method somewhere that looked like a private method got removed. Different issue though.

In general I mostly miss having constructor being the first thing in the class, especially since we are not using stage 2 transformations in Babel to enable member fields, so the only place the members can be declared is in the constructor.
